hw/top_earlgrey/data/top_earlgrey.hjson

```diff
@@ -996,77 +996,289 @@
-    // and there is no way to disable this. DIO inputs on the other hand are by
-    // default not connected.
-    // TODO: need to add mechanism to mark them as wakeup pins.
+    // Number of wakeup detectors to instantiate, and bitwidth for the wakeup counters.
     num_wkup_detect: 8
     wkup_cnt_width:  8
-
-    // Below fields are generated.
-    // inputs: [
-    //  { name: "xxx", width: xx },
-    //  ]
-    // outputs: [
-    //  { name: "xxx", width: xx },
-    //  ]
-    // inouts: [
-    //  { name: "xxx", width: xx },
-    //  ]
-  }
-
-  // TODO: need to overhaul this datastructure.
-  // PADS instantiation
-  //  Number of in/outs and the numer of PAD instances doesn't have to be
-  //  same. The number given below excludes clock/reset and other necessary
-  //  PADS but only defines GPIO pads.
-  padctrl: {
-    attr_default: ["STRONG"],
-    pads: [
-      { name: "ChA" type: "IO_33V", count: 32 }, // Accessing as ChA[0] .. ChA[31]
-      { name: "ChB" type: "IO_33V", count: 4, attr: ["KEEP", "WEAK"]},
-      { name: "ChC" type: "IO_33V", count: 4, attr: ["KEEP", "STRONG"]},
-    ]
   }
 
+  // Implementation targets.
+  // This defines the configuration of the target-specific chip-levels  to
+  // generate from the shared template. Each target uses the same base
+  // configuration for the  pinmux and pinout as defined above, and the
+  // generated software constants for the pinmux DIF do not change among the
+  // implementation targets. However, in order to accommodate slight
+  // differences among the ASIC, FPGA emulation and simulation environments,
+  // it is possible to make very limited pinout changes below. In particular,
+  // it is possible to remove and tie-off specific pads, or add more 'manual'
+  // pads that need to be manually connected in the template. It is also possible
+  // to override the JTAG and strap locations indices, since the testing and DFT
+  // setups may differ among the targets.
+  targets: [
+    { name: 'asic',
+
+      // Pinout Changes.
+      pinout: {
+        // This is a list of pad names to remove from the base pinout defined
+        // above. Removed pads will be stubbed off such that their inputs
+        // signals are driven  with zero. Output signals and output enables
+        // will be left unconnected. If no changes are needed, this list can
+        // be left empty.
+        remove_pads: [],
+
+        // This is a list of additional "manual" pads to add to the pinout.
+        // Pads specified in this list have to be manually connected in the
+        // template and  cannot be muxed, nor can they be referenced in the
+        // pinout config. The pad entries have the same format as the pad
+        // entries in the pinout configuration above, with the difference that
+        // the 'muxed' key must not be specified.
+        add_pads: [],
+      },
+
+      pinmux: {
+        // Special signal positions. Each entry in the list below creates a
```

Comment:
in terms of defining scan roles, would that go here..? or is it elsewhere?

Reply:
I thought about that.
Technically it could go here if we are able to put this into the open source (it would be the right place).
The problem is that this is apparently something that needs to be kept closed, hence we need to work around this with a special package and the partner fusesoc flag.
I would honestly prefer the first method, since the latter is not going to be pretty and easy to maintain...

Reply:
do you think it's possible to some kind of "import cfg" option? That way maybe we can look for the existence of another import that would define these things.  And maybe that would allow us to get away with using the partner flag.

What do you think?

Reply:
I am looking into this next. Adding such an import cfg option would probably work, but I am not sure it makes sense to add even more "partner" select options at a different place in the build flow.
Also, Nuvoton would have to regenerate the toplevel that we currently check in if we go down that route.

Let me try to find a solution that works with the existing partner switches first.